File: CSharp/SmartVote.cs
using Neo.SmartContract.Framework;
using Neo.SmartContract.Framework.Services.Neo;
using Neo.SmartContract.Framework.Services.System;

namespace Neo.SmartContract
{
    public class SmartVote : Framework.SmartContract
    {
        /// <summary>
        /// SmartContract's Entry Point
        /// Parameter list:
        ///     https://docs.neo.org/en-us/sc/Parameter.html
        /// </summary>
        /// <param name="operation">The name of the operation to perform (07)</param>
        /// <param name="args">List of arguments along with the operation(10)</param>
        /// <returns>result of the execution(10)</returns>
        public static object Main(string operation, params object[] args)
        {
            if (operation == "register_proposal") return Register_proposal(args);
            else if (operation == "vote") return Vote(args);
            else if (operation == "count") return Count(args);

            return false;
        }

        /// <summary>
        /// Register a New prosal
        /// </summary>
        /// <param name="args">list of arguments [PROPOSA ID, PROPOSAL TEXT, AUTHORIZED ADDR 1, AUTHORIZED ADDR N, ...]</param>
        /// <returns>Result of the execution</returns>
        static bool Register_proposal(object[] args)
        {
            // check proposal existence
            string proposal_id = (string)args[0];
            StorageContext ctx = Storage.CurrentContext;
            if (Storage.Get(ctx, proposal_id).Length != 0)
            {
                Runtime.Log("Already a Proposal");
                return false;
            }

            // proposal arguments
            object[] proposal = new object[4];
            proposal[0] = args[1];    // description text
            proposal[1] = "0";        // yes counter
            proposal[2] = "0";        // no counter

            // iterate from the third parameter to extract authorized address
            int index = 2;
            string[] proposal_address = new string[args.Length - 2];
            while (index < args.Length)
            {
                string address = (string)args[index];
                // check address format
                if (address.Length != 20)
                {
                    Runtime.Log("bad address format");
                    return false;
                }

                proposal_address[index - 2] = address;
                index += 1;
            }
            proposal[3] = proposal_address; // authorized address

            // serialize proposal and write to storage
            byte[] proposal_storage = proposal.Serialize();
            Storage.Put(ctx, proposal_id, proposal_storage);

            return true;
        }

        /// <summary>
        /// Vote for a proposal
        /// </summary>
        /// <param name="args">list of arguments [PROPOSAL_ID, VOTE]</param>
        /// <returns>result of the execution</returns>
        static bool Vote(object[] args)
        {
            if (args.Length != 2)
                return false;

            // get proposal from storage
            byte[] proposal_id = (byte[])args[0];
            int vote = (int)args[1];

            // Extract caller
            Transaction tx = (Transaction)ExecutionEngine.ScriptContainer;
            TransactionOutput[] inputs = tx.GetReferences();
            TransactionOutput reference = inputs[0];
            byte[] caller = reference.ScriptHash;

            StorageContext ctx = Storage.CurrentContext;
            byte[] proposal_storage = Storage.Get(ctx, proposal_id);

            // check proposal existence
            if (proposal_storage.Length == 0)
            {
                Runtime.Log("Proposal not found");
                return false;
            }

            // check if address already voted
            byte[] key = proposal_id.Concat(caller);
            if (Storage.Get(ctx, key).Length > 0)
            {
                Runtime.Log("Already voted");
                return false;
            }

            // Deserialize proposal array
            object[] proposal = (object[])proposal_storage.Deserialize();

            // check if tx is right signed
            if (!Runtime.CheckWitness(caller))
            {
                Runtime.Log("You are not who you say");
                return false;
            }

            // iterate authorized address
            int index = 0;
            bool authorized = false;
            while (index < ((string[])proposal[3]).Length)
            {
                // for address in proposal[3]:
                if ((object)((string[])proposal[3])[index] == caller)
                {
                    authorized = true;
                    break;
                }
                index += 1;
            }
            if (!authorized)
            {
                Runtime.Log("Not allowed to vote");
                return false;
            }

            // increment vote counter
            if (vote == 1)
            {
                Runtime.Log("Yes!");
                proposal[1] = (int)proposal[1] + 1;
            }
            else
            {
                Runtime.Log("No!");
                proposal[2] = (int)proposal[2] + 1;
            }

            // serialize proposal and write to storage
            proposal_storage = proposal.Serialize();
            Storage.Put(ctx, proposal_id, proposal_storage);

            // mark address as voted
            Storage.Put(ctx, key, 1);
            return true;
        }

        /// <summary>
        /// Count the votes of a proposal
        /// </summary>
        /// <param name="args">list of arguments [PROPOSAL_ID]</param>
        /// <returns>Result of the execution</returns>
        static object Count(object[] args)
        {
            if (args.Length != 1)
                return false;

            // get proposal from storage
            byte[] proposal_id = (byte[])args[0];
            StorageContext ctx = Storage.CurrentContext;
            byte[] proposal_storage = Storage.Get(ctx, proposal_id);

            // check proposal existence
            if (proposal_storage == null || proposal_storage.Length == 0)
            {
                Runtime.Log("Proposal not found");
                return false;
            }

            // Deserialize proposal array
            object[] proposal = (object[])proposal_storage.Deserialize();

            // get proposal description
            Runtime.Log((string)proposal[0]);

            // get proposal votes
            Runtime.Log((string)proposal[1]);
            Runtime.Log((string)proposal[2]);

            return proposal;
        }
    }
}

File: Python/SmartVote.py
"""
SmartVote - smart contract example
===================================

Author: belane
Email: belane@cityofzion.io

Date: 2018

"""

from boa.interop.Neo.Storage import Get, Put
from boa.interop.Neo.Runtime import Serialize, Deserialize
from boa.interop.System.ExecutionEngine import GetScriptContainer
from boa.interop.Neo.Transaction import GetReferences
from boa.interop.Neo.Output import GetScriptHash
from boa.builtins import concat


# get storage context
ctx = GetContext()


def Main(operation, args):
    """
    :param operation: str The name of the operation to perform (07)
    :param args: list of arguments along with the operation (10)
    :return: bool, result of the execution (01)
    doc: docs.neo.org/en-us/sc/tutorial/Parameter.html
    """

    if operation == 'register_proposal':
        return register_proposal(args)

    elif operation == 'vote':
        return vote(args)

    elif operation == 'count':
        return count(args)

    else:
        print("Unknown operation")
        return False


def register_proposal(args):
    """
    Register a New prosal
    :param args: list of arguments [PROPOSA ID, PROPOSAL DESCRIPTION, AUTHORIZED ADDR 1, AUTHORIZED ADDR N, ...]
    :return: bool, result of the execution
    """
    if len(args) < 4:
        return False

    # check proposal existence
    proposal_id = args[0]
    if Get(ctx, proposal_id):
        print("Already a Proposal")
        return False

    # proposal arguments
    proposal = []
    proposal.append(args[1])    # proposal description
    proposal.append(0)          # yes counter
    proposal.append(0)          # no counter

    # iterate from the third parameter to extract authorized addresses
    proposal_address = []
    index = 2
    while index < len(args):

        address = args[index]
        # check address format
        if len(address) != 20:
            print('bad address format')
            return False

        proposal_address.append(address)
        index += 1

    proposal.append(proposal_address) # authorized addresses

    # serialize proposal and write to storage
    proposal_storage = Serialize(proposal)
    Put(ctx, proposal_id, proposal_storage)

    return True

def vote(args):
    """
    Vote for a proposal
    :param args: list of arguments [PROPOSAL_ID, VOTE]
    :return: bool, result of the execution
    """
    if len(args) != 2:
        return False

    # get proposal from storage
    proposal_id = args[0]
    proposal_storage = Get(ctx, proposal_id)

    # check proposal existence
    if not proposal_storage:
        print("Proposal not found")
        return False

    # get caller address
    references = GetScriptContainer().References
    if len(references) < 1:
        return False
    caller_addr = references[0].ScriptHash

    # check if address already voted
    if Get(ctx, concat(proposal_id, caller_addr)):
        print('Already voted')
        return False

    # Deserialize proposal array
    proposal = Deserialize(proposal_storage)

    # iterate authorized address
    index = 0
    while index < len(proposal[3]):
        if proposal[3][index] == caller_addr:
            authorized = True
            break
        index += 1

    if not authorized:
        print('Not allowed to vote')
        return False

    # increment vote counter
    if args[1] == 1:
        print('Yes!')
        proposal[1] = proposal[1] + 1
    else:
        print('No!')
        proposal[2] = proposal[2] + 1

    # serialize proposal and write to storage
    proposal_storage = Serialize(proposal)
    Put(ctx, proposal_id, proposal_storage)

    # mark address as voted
    Put(ctx, concat(proposal_id, caller_addr), True)

    return True

def count(args):
    """
    Count the votes of a proposal
    :param args: list of arguments [PROPOSAL_ID]
    :return: bool, result of the execution
    """
    if len(args) != 1:
        return False

    # get proposal from storage
    proposal_id = args[0]
    proposal_storage = Get(ctx, proposal_id)

    # check proposal existence
    if not proposal_storage:
        print("Proposal not found")
        return False

    # deserialize proposal array
    proposal = Deserialize(proposal_storage)

    # get proposal description
    description = proposal[0]
    print(description)

    # get proposal votes
    votes_yes = proposal[1]
    votes_no = proposal[2]
    print(votes_yes)
    print(votes_no)

    return True
